File: sim/resource.py
import sim.character
import sim.event_loop


class Resource(sim.event_loop.Listener):
    def __init__(self, character: "sim.character.Character", short_rest: bool = False):
        self.num = 0
        self.max = 0
        self.reset_on_short_rest = short_rest
        character.events.add(self, ["short_rest", "long_rest"])

    def increase_max(self, amount: int):
        self.max += amount

    def short_rest(self):
        if self.reset_on_short_rest:
            self.reset()

    def long_rest(self):
        self.reset()

    def reset(self):
        self.num = self.max

    def use(self):
        if self.num > 0:
            self.num -= 1
            return True
        return False

    def has(self):
        return self.num > 0

File: src/sim/resources/Resource.ts
import { log } from "../../util/Log"
import { Character } from "../Character"

export class Resource {
    name: string
    count: number = 0
    max: number = 0

    incrementOnShortRest: boolean = false
    resetOnShortRest: boolean = false
    resetOnLongRest: boolean = false

    constructor(args: {
        name: string
        character: Character
        initialMax?: number
        initialCount?: number
        incrementOnShortRest?: boolean
        resetOnShortRest?: boolean
        resetOnLongRest?: boolean
    }) {
        this.name = args.name
        this.count = args.initialCount ?? args.initialMax ?? 0
        this.max = args.initialMax ?? 0
        args.character.events.on("short_rest", () => this.shortRest())
        args.character.events.on("long_rest", () => this.longRest())
        this.resetOnShortRest = args.resetOnShortRest ?? false
        this.resetOnLongRest = args.resetOnLongRest ?? false
    }

    addMax(amount: number): void {
        this.max += amount
    }

    add(amount: number, overcount?: boolean): void {
        let newCount = this.count + amount
        if (!overcount) {
            newCount = Math.min(this.max, newCount)
        }
        this.count = newCount
    }

    use(amount: number = 1, reason?: string): boolean {
        if (this.count >= amount) {
            log.record(
                `Resource used (${this.name}${reason ? ` (${reason})` : ""})`,
                amount
            )
            this.count -= amount
            return true
        }
        return false
    }

    has(amount: number = 1): boolean {
        return this.count >= amount
    }

    reset(): void {
        this.count = this.max
    }

    private shortRest(): void {
        if (this.incrementOnShortRest) {
            this.add(1)
        }
        if (this.resetOnShortRest) {
            this.reset()
        }
    }

    private longRest(): void {
        if (this.resetOnLongRest) {
            this.reset()
        }
    }
}
